File: Daily/2023/2328_numberOfIncreasingPathsInAGrid.cs
public class Solution {
    int MOD = (int)Math.Pow(10.0, 9.0) + 7;
    int num_rows, num_cols;
    int [,] dp;
    int [,] directions;

    private int dfs(ref int[][] grid, int row, int col) {
        if (dp[row, col] != 0) {
            return dp[row, col];
        }

        dp[row, col] = 1;
        int nrow, ncol;
        for (int dir = 0; dir < 4; dir++) {
            nrow = row + directions[dir, 0];
            ncol = col + directions[dir, 1];

            if (
                nrow < 0 ||
                nrow >= num_rows ||
                ncol < 0 ||
                ncol >= num_cols
            ) {
                continue;
            }

            if (grid[nrow][ncol] >= grid[row][col]) {
                continue;
            }

            dp[row, col] = (dp[row, col] + dfs(ref grid, nrow, ncol)) % MOD;
        } 
    
        return dp[row, col];
    }

    public int CountPaths(int[][] grid) {
        // graph[row][col]: strickly increasing path end at position (row, col)
        num_rows = grid.Length;
        num_cols = grid[0].Length;

        dp = new int [num_rows, num_cols];
        directions = new int [4, 2] {
            {+1, 0},
            {0, +1},
            {-1, 0},
            {0, -1}
        };

        int ret = 0;
        for (int row = 0; row < num_rows; row++) {
            for (int col = 0; col < num_cols; col++) {
                ret = (ret + dfs(ref grid, row, col) ) % MOD;
            }
        }

        return ret;
    }
}

File: Categories/Dynamic_Programming/2328_numberOfIncreasingPathsInAGrid.py
from typing import *
from collections import defaultdict


class Solution:
    def dfs(self, row: int, col: int) -> int:
        if self.dp[row][col] != 0:
            return self.dp[row][col]
        
        self.dp[row][col] = 1
        for direction in self.directions:
            drow, dcol = direction
            nrow, ncol = row + drow, col + dcol

            if (
                nrow < 0 or nrow >= self.num_rows or
                ncol < 0 or ncol >= self.num_cols
            ):
                continue

            if self.grid[nrow][ncol] >= self.grid[row][col]:
                continue
            
            self.dp[row][col] = (self.dp[row][col] + self.dfs(nrow, ncol)) % self.MOD
        
        return self.dp[row][col]


    def countPaths(self, grid: List[List[int]]) -> int:
        self.MOD = 10**9 + 7
        self.grid = grid
        # dp[row][col]: strictly increasing path ending in position (row, col)
        self.dp = defaultdict(lambda : defaultdict(int))
        self.num_rows, self.num_cols = len(grid), len(grid[0])
        self.directions = [
            [+1, 0],
            [0, +1],
            [-1, 0],
            [0, -1],
        ]

        ret = 0
        for row in range(self.num_rows):
            for col in range(self.num_cols):
                ret = (ret + self.dfs(row, col)) % self.MOD
        
        return ret